docs(adk): audit foundations and concept pages (#444)
botpress/docs: adk/concepts/conversations.mdx

## Conversation instance

The `handler` receives a `conversation` object that provides methods to interact with the current conversation. It exposes `conversation.id` (the conversation ID) and `conversation.channel` (the channel name) as read-only properties.

### Loading a conversation by ID

> You can load any conversation by its ID using the static `Conversation.get()` method:

Reviewer: Need to add more details that the input needs to be the conversation ID as a string

Author: yes added!

    Optional Zod schema defining the conversation state structure. Defaults to an empty object if not provided.
  </ResponseField>
  <ResponseField
    name="events"
    type="string[]"
  >
    Optional array of event names this conversation should listen to. Supports integration events with a `conversationId` (e.g., `"webchat:conversationStarted"`) and custom events defined in `agent.config.ts`. By default, conversations only receive messages, not events.
  </ResponseField>
  <ResponseField
    name="chat"
    type="(props: object) => Chat"
  >

>     Optional factory function to provide a custom `Chat` instance for the conversation.

Reviewer: Might be confusing for someone who doesn't know what a `factory function` is. Also need more details on what a `Chat` is so probably linking to the page will be needed here

Author: Good point!

  </ResponseField>
</Expandable>

### Handler parameters

> The handler uses a discriminated union on its function parameters, based on the `type` field.

Reviewer: reword the `discriminated union` ~ explain it simply for someone who won't know what a union is and why its discriminated

Author: removed!

> The handler function receives different parameters depending on the incoming request type. The handler uses a discriminated union based on the `type` field.

Reviewer 2: It's a bit repetitive, could we do something like:

```suggestion
The handler uses a discriminated union on its function parameters, based on the `type` field.
```

    required
  >
    Function to execute autonomous AI logic. Used to run the agent with instructions, tools, knowledge bases, etc.
  </ResponseField>
  <ResponseField
    name="chat"
    type="Chat"
    required
  >
    Chat instance for sending messages and managing the conversation transcript.
  </ResponseField>
  <ResponseField
    name="channel"
    type="string"
    required
  >
    The channel this message or event came from.
  </ResponseField>
</Expandable>

#### Message handler

When `type` is `"message"`:

<Expandable title="Message handler parameters">
  <ResponseField
    name="type"
    type="'message'"
    required
  >
    Indicates this is a message request.
  </ResponseField>
  <ResponseField
    name="message"
    type="object"
    required
  >

>     Message object containing `type` and `payload` properties. Typed based on the channel at runtime.

Reviewer 2: ```suggestion
    Message object containing `type` and `payload` properties. Runtime type is channel dependent.
```

  </ResponseField>
</Expandable>

#### Event handler

When `type` is `"event"`:

<Expandable title="Event handler parameters">
  <ResponseField
    name="type"
    type="'event'"
    required
  >
    Indicates this is an event request.
  </ResponseField>
  <ResponseField
    name="event"
    type="object"
    required
  >

>     Event object containing `type` and `payload` properties from integrations or custom events. Typed based on the channel at runtime.

Reviewer 2: ```suggestion
    Event object containing `type` and `payload` properties from integrations or custom events. Runtime type is channel dependent.
```

The `execute` function accepts the following props:

<Expandable title="Execute props">
  <ResponseField
    name="instructions"
    type="string | (() => string) | (() => Promise&lt;string&gt;)"
    required
  >
    Instructions for the AI agent. Can be a string or a function that returns a string.
  </ResponseField>
  <ResponseField
    name="tools"
    type="object[]"
  >
    Optional array of tools the agent can use.
  </ResponseField>
  <ResponseField
    name="objects"
    type="object[]"
  >
    Optional array of objects the agent can interact with.
  </ResponseField>
  <ResponseField
    name="exits"
    type="Record&lt;string, object&gt;"
  >
    Optional record of exit handlers.
  </ResponseField>
  <ResponseField
    name="signal"
    type="AbortSignal"
  >
    Optional abort signal to cancel execution.
  </ResponseField>
  <ResponseField
    name="hooks"
    type="object"
  >
    Optional hooks for customizing behavior (`onBeforeTool`, `onAfterTool`, `onTrace`, `onBeforeExecution`, `onExit`, `onIterationStart`, `onIterationEnd`).
  </ResponseField>
  <ResponseField
    name="temperature"
    type="number | (() => number) | (() => Promise&lt;number&gt;)"
  >

>     Optional temperature for the AI model.

Reviewer: question: Is there a default?

Author: added the default = 0.7

  </ResponseField>
  <ResponseField
    name="model"
    type="string | string[] | (() => string | string[]) | (() => Promise&lt;string | string[]&gt;)"
  >
    Optional model name(s) to use. Can be a string, array of strings, or a function that returns either.
  </ResponseField>
  <ResponseField
    name="reasoningEffort"
    type="'low' | 'medium' | 'high' | 'dynamic' | 'none'"
  >
    Optional reasoning effort for models that support reasoning. `"none"` disables reasoning, `"dynamic"` lets the provider decide.
  </ResponseField>
  <ResponseField
    name="knowledge"
    type="object[]"
  >
    Optional array of knowledge bases to use.
  </ResponseField>
  <ResponseField
    name="iterations"
    type="number"
  >
    Optional maximum number of iterations (loops). Defaults to 10.
  </ResponseField>
</Expandable>

<Tip>

> The ADK automatically discovers and registers all exported `Conversation` instances in the `src/conversations/` directory. You can use either default or named exports.

Reviewer: question: If we can do named exports, is there an example of how this is done? Currently the code examples only show `default` versions unless I'm not understanding what named exports mean

Author: good point, added an example!
